### ISSUE TITLE
Wrong train-test split for XGB evaluation ?

### ISSUE THREAD

**Issue by BobAubouin** [lorenzkap/MAP_TFT#1]
Dear Lorenzo,

I was reading your paper "Development and external validation of temporal fusion transformer models for continuous intraoperative blood pressure forecasting", and I noticed the high performance of the XGB model in the internal test set that seems unusual.

In the paper, it is written that XGB is trained "_on the same training data set used for the TFT model_". However, when reading the code, both in file XGB.py and XGB_Plot.ipynb, I noticed that although you defined your own "train_test_split" function at [beginning](https://github.com/lorenzkap/MAP_TFT/blob/151cfdb90f8731883ce8454248521516a1ec29e5/XGB.py#L81), the same function is then [imported from sklearn library](https://github.com/lorenzkap/MAP_TFT/blob/151cfdb90f8731883ce8454248521516a1ec29e5/XGB.py#L438C1-L438C53). Particularly, this is the version from sklearn that is used when [splitting training and test](https://github.com/lorenzkap/MAP_TFT/blob/151cfdb90f8731883ce8454248521516a1ec29e5/XGB.py#L450), thus the data is not separated by patient, which leads to data leakage. This might explain the performance of the XGB model on the internal test set.

Is this the exact version of the code that you used to generate the results of the paper? 
If so, could you try to fix this issue and update the paper?
If not, and it appears that you actually used the same data for the TFT model training and test, do you have some insights about the XGB model's performance?

Thanks for your response,
Bob

**Comment by lorenzkap** [comment]
Thank you for your comment.

I noticed the issue and appreciate you pointing it out. Please note that this is not the code used to generate the final results. In the original version, we imported the IDs from the splitting process for the TFT model. I attempted to create a version that runs without this import, but I made an error. I have now corrected the mistake and added an option to load the IDs from a .csv, as well as support for patient group splitting.

The XGB model’s performance: the overall excellent performance of the XGB model is primarily because there are few challenging cases in the dataset. Since the data are continuous, if a patient is not hypotensive, it is very likely they will remain so, and similarly, if a patient is hypotensive, they are likely to remain that way in subsequent time steps.

Please feel free to ask if you need any further details.

Kind regards,
Lorenz

**Comment by BobAubouin** [comment]
Thanks for your explanations.

I still do not understand why the XGB is performing this well and particularly better than the TFT model. If there are only a few challenging cases in the internal test set, the TFT model should also perform well. 

A simple way to measure how "challenging" the internal data is could be to test the ability of a threshold on the mean arterial signal to predict hypotension. I did it on the external test set using the code that you provided for XGB evaluation to have a comparison (I just used a single feature from the "original_feature_value" vector), and I obtained the following ROC curves:

![Image](https://github.com/user-attachments/assets/3c28ddbb-7056-4dd8-8ba3-c1c39a85eb43)

As you can observe, a simple threshold is able to do better than the XGB model on the external test set for forecast distance of 5 and 7 minutes, which should not be possible if the XGB model just learned to predict hypotension based on the last value of MAP.

It would be nice to test this simple classifier on the internal test set. It will measure how "challenging" the prediction of hypotension is on this set. Furthermore, it could be also nice if you were able to run the XGB training and test again, ensuring that there is no data leakage, either by using the code on this git or by pushing the code that you used for the results of the paper (unless there is a reason that the code for the results is not on this git?).

I'm particularly curious about your results, which are very interesting, and that's why I'm looking for more details. Note also that this issue does not question your work with the TFT model, since it still performs better than the threshold on MAP. I think that this comparison even better highlights the performance of your approach.

Kind regards,
Bob

**Comment by BobAubouin** [comment]
Here is the small code that I used to create the ROC figure. It should be added at the end of the XGB_Plot notebook:

```from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

for shift_index in range(1,len(shifts)):
    labels = df_kor[label_names[shift_index]].astype(bool).values
    original_features = df_kor['meanbp'].values
    not_nan_indices = np.logical_not(np.isnan(labels))
    labels = labels[not_nan_indices]
    original_features = original_features[not_nan_indices]
    print(f"Shift index: {shift_index}")
    print(f"Number of samples: {len(labels)}")
    print(f"prevalence of hypotension: {np.sum(labels == 1) / len(labels)}")
    model = LogisticRegression()
    model.fit(original_features.reshape(-1,1), labels)
    prob_pos = model.predict_proba(original_features.reshape(-1,1))[:, 1]
    y_test = labels
    opt_threshold = optimal_youden_threshold(y_test, prob_pos)
    y_pred = prob_pos > opt_threshold
    fpr, tpr, thresholds = roc_curve(y_test, prob_pos)
    auc = roc_auc_score(y_test, prob_pos)
    plt.plot(fpr, tpr, label=f"Shift: {shifts[shift_index]/-4} minutes (AUC = {auc:.2f})")
plt.xlabel('False Positive Rate')
plt.ylabel('True Positive Rate')
plt.title('ROC Curve of a threshold on MAP on external set')
plt.legend()
plt.grid()
plt.show()```

**Comment by lorenzkap** [comment]
Hi, thanks for your feedback!

I’ve outlined several reasons in the paper. 
My interpretation is that the XGB models seem to perform better in the internal validation because they’re solving a simpler task—a binary (Yes/No) decision at a specific point in time—compared to making continuous predictions. This setup might even allow them to anticipate a clinician’s response, which, in turn, reduces their performance during external validation. Additionally, I suspect that the poorer performance of the XGB model in the external test set is linked to the absence of bolus medication data, which has a high feature importance.

I’ll re-run the analysis using the code available on GitHub to see if this version produces the same results. However, please bear with me as I am currently busy writing my dissertation and working with limited resources, so this may take some time.

Best regards,
Lorenz

**Comment by BobAubouin** [comment]
Hi,

Thanks again for this quick reply !
I agree with your reasons, although I'm not sure that the change of task explain the differences of performances. As you state in your paper, the regression task should allow a better generalization of the model and avoid overfitting. Despite coming from the same data center, the internal test is different from training and I believe that TFT model should outperform the XGB even in the internal setup.

Moreover, the performances of XGB are really high, in the paper the ROC curve for forecast distance of 7 minutes is even slightly higher than the one of three minutes forecast which really seems contradictory.

To verify your hypothesis about the unmeasured bolus in vitalDB data, you can evaluate your model on cases without any bolus (or at least negligible bolus volume).

Also, if your model manages to learn the clinician response, it might be an issue since it will predict a reaction of the care team, and thus predict no hypotension, rather than rising an alert about a needing reaction. I don't think that's what we want at the end.

Good luck with your dissertation ! (But check your results before writing the conclusion ;) )

Kind regards,
Bob
